src/he/pmtud.c

```diff
@@ -0,0 +1,301 @@
+        return he_internal_pmtud_search_completed(conn);
+      }
+      break;
+    case HE_PMTUD_STATE_SEARCH_COMPLETE:
+      // Black hole detected
+      return he_internal_pmtud_blackhole_detected(conn);
+    default:
+      // Do nothing in DISABLED / ERROR states
+      break;
+  }
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_start_base_probing(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_DISABLED:
+    case HE_PMTUD_STATE_SEARCHING:
+    case HE_PMTUD_STATE_SEARCH_COMPLETE:
+      // Valid states
+      break;
+    case HE_PMTUD_STATE_BASE:
+      // Already in base state, do nothing
+      return HE_SUCCESS;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_BASE);
+
+  // Initialize PMTUD internal state
+  conn->pmtud_base = MIN_PLPMTU;
+  conn->pmtud_probe_count = 0;
+
+  // Start probing base mtu
+  return he_internal_pmtud_send_probe(conn, conn->pmtud_base);
+}
+
+he_return_code_t he_internal_pmtud_base_confirmed(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+    case HE_PMTUD_STATE_ERROR:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change to Searching state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_SEARCHING);
+
+  // Start searching
+  conn->pmtud_probe_count = 0;
+  conn->pmtud_is_using_big_step = true;
+  uint16_t probe_size = conn->pmtud_base + PMTUD_PROBE_BIG_STEP;
+
+  return he_internal_pmtud_send_probe(conn, probe_size);
+}
+
+he_return_code_t he_internal_pmtud_confirm_base_failed(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_ERROR);
+
+  // TODO: recheck base pmtu periodically
+  conn->pmtud_probe_count = 0;
+
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_search_completed(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_SEARCHING:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Set the effective pmtu
+  conn->effective_pmtu = MIN(conn->pmtud_probing_size, MAX_PLPMTU);
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_SEARCH_COMPLETE);
+
+  // TODO: stay in this state and check for black hole regularly
+  conn->pmtud_probe_count = 0;
```

Comment:
We may need to reset the `conn->pmtud_is_using_big_step` to true, cause the last search might end up in small_step.
If not, the next search will start with small_step ?

Reply:
We only need to reset this variable `pmtud_is_using_big_step` when entering SEARCHING state. It's currently handled in `he_internal_pmtud_base_confirmed` function.

We'll also reset it here when we implement above "TODO" later.

Reply:
Oh right, the next search will go through base and it will reset it. 
Thanks for clarifying